README.txt
= activeworlds-ffi

== Description

A Ruby/FFI wrapper for the Active Worlds, Inc. SDK providing for the rapid creation and deployment of "bots".

== Requirements

You must have the Active Worlds SDK (currently only the linux version, build 77), and the ffi gem version 0.3.5 installed. The SDK must be on your LD_LIBRARY_PATH. This is a 32-bit only library because of the AW SDK, and so for linking purposes you must have 32-bit versions of all libraries used, in particular libstdc++.so.6 (if you have a different c++ lib, change the name in lib/activeworlds_ffi.rb).

== Examples

See the examples directory in the gem to see full examples. A simple activity tracking bot is provided there. It is based on a nice support library that handles a number of events such as disconnects, and asynchronous creation, login, and entrance of batches of bots.

For ease though, here is a translation of the Active Worlds sample hello bot.

#! /usr/bin/env ruby19
require 'rubygems'
require 'activeworlds_ffi'

include ActiveworldsFFI

def handle_avatar_add
  s = "Hello #{aw_string(AW_AVATAR_NAME)}! I'm the broken record."
  # we're using a console message instead of aw_say(s) since that might be annoying
  aw_string_set(AW_CONSOLE_MESSAGE, s)
  aw_console_message(aw_int(AW_AVATAR_SESSION))
  puts(s)
end

if ARGV.nil? || ARGV.size < 5
  puts("Usage: hello.rb server port citizen_id privilege_password world")
  exit(1)
end

# initialize Active Worlds API
rc = aw_init(AW_BUILD)
if( rc != RC_SUCCESS )
  printf("Unable to initialize API (reason %d)\n", rc)
  exit( 1 ) 
end

# assign the proc to a constant so that it never gets garbage collected
AVATAR_ADD_HANDLER = Proc.new { handle_avatar_add }

# install handler for avatar_add event 
aw_event_set(AW_EVENT_AVATAR_ADD, AVATAR_ADD_HANDLER)
  
# create bot instance 
rc = aw_create(ARGV[0], ARGV[1].to_i, nil);
if rc != RC_SUCCESS
  printf "Unable to create bot instance (reason %d)\n", rc
  exit 1
end
  
# log bot into the universe 
aw_int_set AW_LOGIN_OWNER, ARGV[2].to_i
aw_string_set AW_LOGIN_PRIVILEGE_PASSWORD, ARGV[3]
aw_string_set AW_LOGIN_APPLICATION, "SDK Sample Application #1"
aw_string_set AW_LOGIN_NAME, "GreeterBot"
rc = aw_login
if rc != RC_SUCCESS
  printf("Unable to login (reason %d)\n", rc)
  exit(1)
end
  
# log bot into the world named on the command line 
rc = aw_enter (ARGV[4]);
if (rc != RC_SUCCESS)
  printf("Unable to enter world (reason %d)\n", rc)
  exit(1)
end
  
# announce our position in the world 
aw_int_set(AW_MY_X, 1000) #/* 1W */
aw_int_set(AW_MY_Z, 1000) #/* 1N */
aw_int_set(AW_MY_YAW, 2250)  #/* face towards GZ */
rc = aw_state_change 
if rc != RC_SUCCESS
  printf "Unable to change state (reason %d)\n", rc
  exit 1
end
  
#  /* main event loop */
while aw_wait(-1) == RC_SUCCESS
end
  
#  /* close everything down */
aw_destroy
aw_term



History.txt

== 4.2.77.1 (2009-08-08)

* This is the initial version of the Active Worlds version 4.2 build 77 ffi based wrapper.

examples/tracker.rb

require 'base_bot'

$BOTDIR = File.dirname(__FILE__)
# record the pid
File.open(File.join($BOTDIR, 'pids','tracker.pid'),'w') {|f| f.write(Process.pid) }

class ActivityLogging < BaseBot
    
  def initialize(world_name)
    super(world_name)
  end
  
  def log_avatar_add
    set_event_handler :AW_EVENT_AVATAR_ADD, Proc.new {
      a = attributes_for(:AW_EVENT_AVATAR_ADD)
      info "ENTER | #{a[:AW_AVATAR_NAME]} | #{a[:AW_AVATAR_SESSION]}" +
          " | #{a[:AW_AVATAR_X]} | #{a[:AW_AVATAR_Y]} | #{a[:AW_AVATAR_Z]}" +
          " | #{a[:AW_AVATAR_YAW]}"
    }
  end
  
  def log_chat
    set_event_handler :AW_EVENT_CHAT, Proc.new {
      a = attributes_for(:AW_EVENT_CHAT)
      info "CHAT | #{a[:AW_AVATAR_NAME]} | #{a[:AW_CHAT_MESSAGE]}"
    }
  end
  
  def log_avatar_change
    set_event_handler :AW_EVENT_AVATAR_CHANGE, Proc.new {
      a = attributes_for(:AW_EVENT_AVATAR_CHANGE)
      info "MOVED | #{a[:AW_AVATAR_NAME]} | #{a[:AW_AVATAR_SESSION]}" +
        " | #{a[:AW_AVATAR_X]} | #{a[:AW_AVATAR_Y]} | #{a[:AW_AVATAR_Z]}" +
        " | #{a[:AW_AVATAR_YAW]}"
    }
  end

  def log_avatar_delete
    set_event_handler :AW_EVENT_AVATAR_DELETE, Proc.new {
      a = attributes_for(:AW_EVENT_AVATAR_DELETE)
      info "LEFT | #{a[:AW_AVATAR_NAME]} | #{a[:AW_AVATAR_SESSION]}"
    }
  end

  def log_avatar_click
    set_event_handler :AW_EVENT_AVATAR_CLICK, Proc.new {
      a = attributes_for(:AW_EVENT_AVATAR_CLICK)
      info "AVATAR_CLICK | #{a[:AW_AVATAR_NAME]} | #{a[:AW_CLICKED_NAME]}"
    }
  end
  
  def log_object_click
    set_event_handler :AW_EVENT_OBJECT_CLICK, Proc.new {
      a = attributes_for(:AW_EVENT_OBJECT_CLICK)
      info "OBJECT_CLICK | #{a[:AW_AVATAR_NAME]} | #{a[:AW_CELL_X]} | #{a[:AW_CELL_Z]}" +
        " | #{a[:AW_OBJECT_ID]} | #{a[:AW_OBJECT_TYPE]} | #{a[:AW_OBJECT_X]} | #{a[:AW_OBJECT_Y]}" +
        " | #{a[:AW_OBJECT_Z]} | #{a[:AW_OBJECT_MODEL]} | #{a[:AW_OBJECT_DESCRIPTION]} | #{a[:AW_OBJECT_ACTION]}"
    }
  end

  def log_url_click
    set_event_handler :AW_EVENT_URL_CLICK, Proc.new {
      a = attributes_for(:AW_EVENT_URL_CLICK)
      info "URL_CLICK | #{a[:AW_AVATAR_NAME]} | #{a[:AW_AVATAR_SESSION]} | #{a[:AW_URL_NAME]}"
    }
  end

  def self.for_world(world_name)
    bot = super( world_name )
    bot.log_avatar_add
    bot.log_chat
    bot.log_avatar_change
    bot.log_avatar_delete
    bot.log_avatar_click
    bot.log_object_click
    bot.log_url_click
    bot
  end
  
  def self.do_work
    @@bots[:instances].each do |bot|
      bot.set_self_as_instance
    end
  end
  
end

def run(options, logger)
  ActivityLogging.configure(options, logger)
  worlds = YAML.load_file('worlds.yml')
  ActivityLogging.start_worlds_by_batch( ActivityLogging.config_worlds.collect {|world_group| worlds[world_group]}.flatten )
  while 0 == ActivityLogging.wait(10000)
    #put non-event work here
  end
rescue => ex
  $ActivityLogger.fatal("#{ex.class} : #{ex.message}\n#{ex.backtrace.join("\n")}")
end

config = YAML.load_file("bots_config.yml")["tracker"]
$ActivityLogger = BotLogger.new(config)
run(config, $ActivityLogger)



examples/bot_template.rb
# This is an example template
#
#

require 'base_bot'

$BOTDIR = File.dirname(__FILE__)
# record the pid
File.open(File.join($BOTDIR, 'pids','BOT.pid'),'w') {|f| f.write(Process.pid) }

class BOT < BaseBot
  
  def initialize(world_name)
    super(world_name)
    
  end
  
  def log_avatar_add
    set_event_handler :AW_EVENT_AVATAR_ADD, Proc.new {
      a = attributes_for(:AW_EVENT_AVATAR_ADD)
      
    }
  end

  # An example for how to setup 
  def handle_hud_callback_result
    set_callback_handler :AW_CALLBACK_HUD_RESULT, Proc.new {|rc|
      a = attributes_for(:AW_CALLBACK_HUD_RESULT)
      debug "HUD_RESULT | #{rc_sym(rc)} | #{a[:AW_HUD_ELEMENT_SESSION]} | #{a[:AW_HUD_ELEMENT_ID]}"
    }
  end
  
  # called by start_worlds_by_batch
  def self.for_world(world_name)
    bot = super( world_name )
    # register callbacks
    
    bot
  end
  
  def self.do_work
    @@bots[:instances].each do |bot|
      bot.set_self_as_instance
    end
  end
end

def run(options, logger)
  BOT.configure(options, logger)
  worlds = YAML.load_file('worlds.yml')
  BOT.start_worlds_by_batch( BOT.config_worlds.collect {|world_group| worlds[world_group]}.flatten )
  while 0 == BOT.wait(1000)
    #put non-event work here
  end
rescue => ex
  $BOTLogger.fatal("#{ex.class} : #{ex.message}\n#{ex.backtrace.join("\n")}")
end

config = YAML.load_file("bots_config.yml")["BOT"]
$BOTLogger = BotLogger.new(config)
run(config, $BOTLogger)


examples/hello.rb
#! /usr/bin/env ruby19
require 'rubygems'
require 'activeworlds_ffi'

include ActiveworldsFFI

def handle_avatar_add
  s = "Hello #{aw_string(AW_AVATAR_NAME)}! I'm the broken record."
  # we're using a console message instead of aw_say(s) since that might be annoying
  aw_string_set(AW_CONSOLE_MESSAGE, s)
  aw_console_message(aw_int(AW_AVATAR_SESSION))
  puts(s)
end

if ARGV.nil? || ARGV.size < 3
  puts("Usage: hello.rb citizen_id privilege_password world")
  exit(1)
end

# initialize Active Worlds API
rc = aw_init(AW_BUILD)
if( rc != RC_SUCCESS )
  printf("Unable to initialize API (reason %d)\n", rc)
  exit( 1 ) 
end

# assign the proc to a constant so that it never gets garbage collected
AVATAR_ADD_HANDLER = Proc.new { handle_avatar_add }

# install handler for avatar_add event 
aw_event_set(AW_EVENT_AVATAR_ADD, AVATAR_ADD_HANDLER)
  
# create bot instance 
rc = aw_create("atlantis.activeworlds.com", 5870, nil);
if rc != RC_SUCCESS
  printf "Unable to create bot instance (reason %d)\n", rc
  exit 1
end
  
# log bot into the universe 
aw_int_set AW_LOGIN_OWNER, ARGV[0].to_i
aw_string_set AW_LOGIN_PRIVILEGE_PASSWORD, ARGV[1]
aw_string_set AW_LOGIN_APPLICATION, "SDK Sample Application #1"
aw_string_set AW_LOGIN_NAME, "GreeterBot"
rc = aw_login
if rc != RC_SUCCESS
  printf("Unable to login (reason %d)\n", rc)
  exit(1)
end
  
# log bot into the world named on the command line 
rc = aw_enter (ARGV[2]);
if (rc != RC_SUCCESS)
  printf("Unable to enter world (reason %d)\n", rc)
  exit(1)
end
  
# announce our position in the world 
aw_int_set(AW_MY_X, 1000) #/* 1W */
aw_int_set(AW_MY_Z, 1000) #/* 1N */
aw_int_set(AW_MY_YAW, 2250)  #/* face towards GZ */
rc = aw_state_change 
if rc != RC_SUCCESS
  printf "Unable to change state (reason %d)\n", rc
  exit 1
end
  
#  /* main event loop */
while aw_wait(-1) == RC_SUCCESS
end
  
#  /* close everything down */
aw_destroy
aw_term

